### PR TITLE
Add React Native Support

### PR DESCRIPTION
**Pull request by lelandrichardson**
to: @lencioni @trotzig 

This PR is broken into 3 commits.  

The first commit is `Add RN Runner App`, which is mostly just the output of `react-native init HappoRunner` to create a standard RN app. The app is only tweaked in a couple of small ways to allow for the snapshotting. This commit could mostly be ignored.

The second commit is `Add RN Target` which adds the actual happo target plugin. The plugin does the following:

1. Launch a websockets server for communication with the app
2. Launch the RN packager with the appropriate configuration
3. Launch the simulator/emulator using Appium. The actual app that is being run is a small RN app that will be included with this module, but it will be running the user's JavaScript RN code.
4. Waits for the app to send it a list of all the snapshots.
5. Sends a message to the app telling it to render a certain snapshot, waiting for the app to respond back with a success message, including a base64 encoded image.
6. Once all images are complete, happo will compare them all and send back results to the happo CLI.

Tested with both iOS and Android.

The third commit is another "example" project that shows the example in action.

### CODE REVIEW

**Comment by lelandrichardson** [comment]
cc @spikebrehm

File: .gitignore [hunk 1]
```diff
@@ -5,3 +5,47 @@ packages/*/public/*.bundle.*
...
+
+# OSX
+#
+.DS_Store
+
+# Xcode
+#
+build/
+*.pbxuser
+!default.pbxuser
```
Comment (trotzig) [inline .gitignore]: `!default.*`?

File: packages/happo-core/src/getImageFromDataURI.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
+const parseDataUri = require('parse-data-uri');
+const { PNG } = require('pngjs');
+
+function getImageFromDataURI(uri) {
+  return new Promise((resolve, reject) => {
+    const img = parseDataUri(uri);
+    const png = new PNG();
+    png.parse(img.data, (err) => {
+      if (err) reject(err);
```
Comment (trotzig) [inline packages/happo-core/src/getImageFromDataURI.js]: Does eslint yell at you for this? I thought we required curlies.

File: packages/happo-core/src/index.js [hunk 3]
```diff
@@ -14,6 +15,7 @@ module.exports = {
   areImagesEqual,
   getImageFromStream,
   getImageFromPath,
+  getImageFromDataURI,
```
Comment (trotzig) [inline packages/happo-core/src/index.js]: Being able to span multiple packages in one PR like this is great!

File: packages/happo-target-react-native/README.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+export S3_SECRET_ACCESS_KEY=<YOUR_SECRET_ACCESS_KEY_VALUE>
+export S3_BUCKET_NAME=<YOUR_BUCKET_NAME>
+
+happo upload
+```
+
+or by adding them in the beginning of the command:
+
+```sh
+S3_ACCESS_KEY_ID=<...> S3_SECRET_ACCESS_KEY=<...> ... happo upload
```
Comment (trotzig) [inline packages/happo-target-react-native/README.md]: Since you direct people at installing `happo` locally (`--save-dev`), maybe it's a good idea to have this say `$(npm bin)/happo upload` instead? We're probably doing the same thing elsewhere though, so I don't know how important it is.

File: packages/happo-target-react-native/src/defaultOptions.js [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+  port: 5000, // websockets port
+  packagerCommand: 'node',
+  packageArgs: [
+    './node_modules/react-native/local-cli/cli.js',
+    'start',
+  ],
+  packagerCwd: process.cwd(),
+  packagerEnv: process.env,
+
+  appiumPath: './node_modules/.bin/appium',
```
Comment (trotzig) [inline packages/happo-target-react-native/src/defaultOptions.js]: I remember you asking about this in Messenger. I take it you put it in configuration to allow people to override. Can you explain scenarios when people would want to do that?

File: packages/happo-target-react-native/src/initializeDriver.js [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    });
+    process.on('exit', (e) => {
+      console.log(e);
+      console.log('[Appium]: Exiting...');
+      appiumProc.kill();
+    });
+  })
+);
+
+let driver;
```
Comment (trotzig) [inline packages/happo-target-react-native/src/initializeDriver.js]: Do you need to reassign this? It looks like you only use it to resolve a promise, so maybe it can be inline in the method?

File: packages/happo-target-react-native/src/native/HappoJob.js [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+      story: getStory(name),
+    });
+  }
+  onSnapshotCompleted(name, uri) {
+    this.socket.emit('snapshotComplete', { name, uri });
+  }
+  render() {
+    const { story } = this.state;
+    return (
+      <View style={{ paddingTop: 20 }}>
```
Comment (trotzig) [inline packages/happo-target-react-native/src/native/HappoJob.js]: Padding still needed? I remember you saying something about this in the previous PR, but I can't recall what you said.

File: packages/happo-target-react-native/src/runVisualDiffs.js [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+          .on('finish', () => {
+            resolve({
+              result: 'new',
+              height: snapshotImage.height,
+            });
+          });
+      });
+    }
+  });
+}
```
Comment (trotzig) [inline packages/happo-target-react-native/src/runVisualDiffs.js]: This method looks like it might be the same as the one used for the firefox target. Perhaps we should move this into happo-core?

File: packages/happo-core/src/getImageFromDataURI.js (hunk 2 above)
Comment (lelandrichardson) [inline packages/happo-core/src/getImageFromDataURI.js]: The Airbnb styleguide [only requires braces with multi-line blocks](https://github.com/airbnb/javascript#blocks)

File: packages/happo-target-react-native/src/runVisualDiffs.js (hunk 8 above)
Comment (lelandrichardson) [inline packages/happo-target-react-native/src/runVisualDiffs.js]: the method is different, although perhaps it could be abstracted into a couple of different functions and then pulled out from there. I tried to keep a balance of not changing *too* much code for this, but can totally try to do this if you think it's worth it.

File: packages/happo-target-react-native/src/native/HappoJob.js (hunk 7 above)
Comment (lelandrichardson) [inline packages/happo-target-react-native/src/native/HappoJob.js]: honestly I still haven't checked to see if the snapshotting method i'm employing will end up grabbing the status bar or not, which is why i'm doing this.

File: packages/happo-target-react-native/src/initializeDriver.js (hunk 6 above)
Comment (lelandrichardson) [inline packages/happo-target-react-native/src/initializeDriver.js]: yeah good point. I had copied this code from our RN code base and tweaked, and doesn't look like it needs to be hoisted anymore.

File: packages/happo-target-react-native/src/defaultOptions.js (hunk 5 above)
Comment (lelandrichardson) [inline packages/happo-target-react-native/src/defaultOptions.js]: yeah i guess i just made it configurable and called that "good enough", but i think I should see if using `npm bin` will get me what i need or not, which may be a more robust way to do this and eliminate it as a config option

File: .gitignore (hunk 1 above)
Comment (lelandrichardson) [inline .gitignore]: i have no idea, TBH. I think this is just github's default "ios project" .gitignore template

File: packages/happo-core/src/index.js (hunk 3 above)
Comment (lelandrichardson) [inline packages/happo-core/src/index.js]: yep!

File: packages/happo-target-react-native/README.md (hunk 4 above)
Comment (lelandrichardson) [inline packages/happo-target-react-native/README.md]: i'll defer to whatever you and joe think is best here! do you want me to change it to `$(npm bin)/happo upload`?

Comment (lelandrichardson) [inline packages/happo-target-react-native/README.md]: also, i just realized this readme is totally wrong and i forgot to change it after i copy-pasted!

File: packages/happo-target-react-native/src/initializeDriver.js [hunk 9]
```diff
@@ -1,13 +1,15 @@
...
 
 const startAppium = ({
-  appiumPath,
   appiumCwd,
   appiumEnv,
   appiumServerPort,
 }) => (
   new Promise((resolve) => {
+    const nodeBin = childProcess.execSync('npm bin');
+    const appiumPath = path.join(nodeBin.toString('utf8').trim(), 'appium');
```
Comment (trotzig) [inline packages/happo-target-react-native/src/initializeDriver.js]: 👍

File: packages/happo-core/src/getImageFromDataURI.js (hunk 2 above)
Comment (trotzig) [inline packages/happo-core/src/getImageFromDataURI.js]: Cool, we should adopt this as well in our org I think.

File: packages/happo-target-react-native/README.md (hunk 4 above)
Comment (trotzig) [inline packages/happo-target-react-native/README.md]: I think that's okay. We could also say `npm install -g happo` for installation instructions, and then use `happo` without prefix throughout the readme. I don't know what's better tbh.

File: packages/happo-target-react-native/src/native/HappoJob.js (hunk 7 above)
Comment (trotzig) [inline packages/happo-target-react-native/src/native/HappoJob.js]: okay, cool.

File: packages/happo-target-react-native/src/runVisualDiffs.js (hunk 8 above)
Comment (trotzig) [inline packages/happo-target-react-native/src/runVisualDiffs.js]: I understand. It would feel a little better if the filesystem was controlled inside happo-core, since it relies on the structure when generating the review page and when results are uploaded.

**Comment by lencioni** [comment]
We regularly use tools like `git log`, `git blame`, and `git show` to understand why a particular line of code is the way it is. To help out our future selves, it would be really great if you could capture the "why" in each of your commit messages. More info: https://brigade.engineering/the-secrets-to-great-commit-messages-106fc0a92a25

File: packages/happo-target-react-native/native.js [hunk 10]
```diff
@@ -0,0 +1,3 @@
+/* eslint-disable */
+// this file isn't compiled, so we just ignore eslint here
```
Comment (lencioni) [inline packages/happo-target-react-native/native.js]: If it isn't too hard to only disable the rules that don't make sense for this file instead of disabling everything, that would be best.

File: packages/happo-target-react-native/scripts/build-runner-android.sh [hunk 11]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env bash
+
+echo "Building HappoRunner.apk"
+node ./node_modules/react-native/local-cli/cli.js run-android --root ./runner
```
Comment (lencioni) [inline packages/happo-target-react-native/scripts/build-runner-android.sh]: I worry a bit about the hard-coded node_modules path in these scripts. Do these need to be shell scripts or can they be fully contained in package.json scripts?

File: packages/happo-target-react-native/src/defaultOptions.js [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+
+const iosAppPath = path.join(
+  __dirname, // path of this file
+  '../', // root of this module
+  'runner/ios/build/Build/Products/Debug-iphonesimulator/HappoRunner.app',
+);
+
+const androidAppPath = path.join(
+  __dirname, // path of this file
+  '../', // root of this module
```
Comment (lencioni) [inline packages/happo-target-react-native/src/defaultOptions.js]: Do you need/want the trailing slash here when using `path.join`?

Also, I worry a little about using __dirname because this may unexpectedly break if this file moves. Probably okay if we have test coverage for this, but if we can find a less brittle way to express this, that might be a good idea.

File: packages/happo-target-react-native/src/defaultOptions.js [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+  'runner/android/app/build/outputs/apk/app-debug.apk',
```
Comment (lencioni) [inline packages/happo-target-react-native/src/defaultOptions.js]: Similarly, each of these directories should be separate args for max cross-platform compatibility.

File: packages/happo-target-react-native/src/defaultOptions.js [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+    deviceName: 'iPhone 6',
+    deviceOrientation: 'portrait',
+  },
+  android: {
+    runnerAppPath: androidAppPath,
+    platformName: 'Android',
+    platformVersion: '6.0',
+    deviceName: 'Android Emulator',
+    deviceOrientation: 'portrait',
+  },
```
Comment (lencioni) [inline packages/happo-target-react-native/src/defaultOptions.js]: These defaults might be good to document in the readme.

File: packages/happo-target-react-native/src/defaultOptions.js [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+};
+
+// non-overridable capabilities
+const platformCaps = {
```
Comment (lencioni) [inline packages/happo-target-react-native/src/defaultOptions.js]: What does "caps" mean? Is it "capabilities"? If so, please use the full word.

File: packages/happo-target-react-native/src/defaultOptions.js [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+  packagerEnv: process.env,
+
+  appiumPath: './node_modules/.bin/appium',
+  appiumCwd: process.cwd(),
+  appiumEnv: process.env,
+  appiumServerHost: 'localhost',
+  appiumServerPort: 4734,
+
+  // caps
+  appiumVersion: '1.6.3',
```
Comment (lencioni) [inline packages/happo-target-react-native/src/defaultOptions.js]: Does this need to stay in sync with the installed version of appium? Why not read it from appium's package.json file instead?

File: packages/happo-target-react-native/src/initializeDriver.js [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+  appiumServerPort,
+}) => (
+  new Promise((resolve) => {
+    const args = [
+      '-p',
+      String(appiumServerPort),
+    ];
+    const appiumProc = childProcess.spawn(appiumPath, args, { cwd: appiumCwd, env: appiumEnv });
+
+    let loadedAppium = null;
```
Comment (lencioni) [inline packages/happo-target-react-native/src/initializeDriver.js]: Should this start as false?

File: packages/happo-target-react-native/src/initializeDriver.js [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+    };
+
+    console.log('[DRIVER]: init');
+
+    driver = wd.promiseChainRemote(serverConfig);
+
+    driver.on('status', (info) => {
+      console.log(info);
+    });
+    driver.on('command', (meth, commandPath, data) => {
```
Comment (lencioni) [inline packages/happo-target-react-native/src/initializeDriver.js]: Similarly, please prefer full words ov abbrvd wds

File: packages/happo-target-react-native/src/initializePackager.js [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+    });
+    process.on('error', (e) => {
+      console.log('error', e);
+      packagerProcess.kill();
+      process.exit(1);
+    });
+    packagerProcess.on('exit', () => {
+      // TODO(lmr): this is a race condition...
+      // this can happen after resolve and it will still break the process. Not sure what the
+      // best way to solve this would be.
```
Comment (lencioni) [inline packages/happo-target-react-native/src/initializePackager.js]: It would be nice to open an issue to track this, and include a link here to the issue.

File: packages/happo-target-react-native/src/native/StoryRenderer.js [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+      format: 'png',
+      quality: 1.0,
+      result: 'data-uri',
+    })
+      .then((uri) => {
+        this.props.onSnapshotCompleted(this.props.name, uri);
+      })
+      .catch((err) => {
+        console.error(err);
+        // TODO(lmr): this shouldn't happen, but we should expose something to user maybe?
```
Comment (lencioni) [inline packages/happo-target-react-native/src/native/StoryRenderer.js]: todo

File: packages/happo-target-react-native/src/native/index.js [hunk 21]
```diff
@@ -0,0 +1,13 @@
+// this is needed to support socket.io, and needs to get called _before_ socket.io gets required
+window.navigator.userAgent = 'react-native';
+
+// still not really sure why we need this... but we do (at least on my machine)...
+// require('./missingBabelHelpers');
```
Comment (lencioni) [inline packages/happo-target-react-native/src/native/index.js]: Can this line and comment be removed?

File: packages/happo-example-react-native-project/package.json [hunk 22]
```diff
@@ -0,0 +1,21 @@
...
+  "dependencies": {
+    "react": "15.4.1",
+    "react-native": "^0.40.0",
+    "happo": "../happo",
+    "happo-viewer": "../happo-viewer",
+    "happo-core": "../happo-core",
+    "happo-target-react-native": "../happo-target-react-native",
+    "happo-uploader-s3": "../happo-target-react-native"
+  }
+}
```
Comment (lencioni) [inline packages/happo-example-react-native-project/package.json]: This package should probably have a basic readme.

File: packages/happo-target-react-native/README.md [hunk 23]
```diff
@@ -7,60 +7,117 @@ If you haven't already installed `happo` you will need to do that:
...
+The entry file should import the client library and define "Stories"
 
-```sh
-export S3_ACCESS_KEY_ID=<YOUR_ACCESS_KEY_VALUE>
-export S3_SECRET_ACCESS_KEY=<YOUR_SECRET_ACCESS_KEY_VALUE>
-export S3_BUCKET_NAME=<YOUR_BUCKET_NAME>
-
-happo upload
-```
+```js
```
Comment (lencioni) [inline packages/happo-target-react-native/README.md]: jsx for better highlighting

File: packages/happo-target-react-native/src/defaultOptions.js (hunk 12 above)
Comment (lelandrichardson) [inline packages/happo-target-react-native/src/defaultOptions.js]: Totally agree about not using `__dirname`. If you have a better idea of how to do this, can you let me know?  I couldn't think of any...

File: packages/happo-target-react-native/README.md [hunk 24]
```diff
@@ -7,60 +7,117 @@ If you haven't already installed `happo` you will need to do that:
...
-Happo uses to access the bucket named `S3_BUCKET_NAME`.
 
-`S3_BUCKET_PATH` can be set as an environment variable to specify a directory
-path for where you want diff images uploaded within the S3 bucket.
+## Defining examples
 
-Furthermore, `S3_REGION` controls what
-[region](http://docs.aws.amazon.com/general/latest/gr/rande.html) is used to
-find or create the bucket.
+Happo expects you to have a `happo.js` entry file at the root that your packager is
```
Comment (lencioni) [inline packages/happo-target-react-native/README.md]: Is this file supposed to be `.happo.js`?

Comment (lelandrichardson) [inline packages/happo-target-react-native/README.md]: no. the RN packager will look for a "happo" entry file.

File: packages/happo-target-react-native/scripts/build-runner-android.sh (hunk 11 above)
Comment (lelandrichardson) [inline packages/happo-target-react-native/scripts/build-runner-android.sh]: i did it for the ios one mainly because i was doing some `cd foo` and `cd -` in it... which is uglier in a package.json script. I don't like making package.json scripts too complicated as they get hard to parse real quick

File: packages/happo-target-react-native/src/defaultOptions.js (hunk 13 above)
Comment (lelandrichardson) [inline packages/happo-target-react-native/src/defaultOptions.js]: good thinking.

File: packages/happo-target-react-native/src/native/StoryRenderer.js (hunk 20 above)
Comment (lelandrichardson) [inline packages/happo-target-react-native/src/native/StoryRenderer.js]: i guess the todo here is kind of redundant. `console.error` in RN will actually expose something to the user. Though maybe we could do more and send some info back to the server and console.log it back in the server process so it shows up in CI logs

File: packages/happo-target-react-native/src/native/index.js (hunk 21 above)
Comment (lelandrichardson) [inline packages/happo-target-react-native/src/native/index.js]: yeah

File: packages/happo-target-react-native/src/defaultOptions.js (hunk 14 above)
Comment (lelandrichardson) [inline packages/happo-target-react-native/src/defaultOptions.js]: better i think is to actually just remove them entirely and make them required

File: packages/happo-target-react-native/src/defaultOptions.js (hunk 16 above)
Comment (lelandrichardson) [inline packages/happo-target-react-native/src/defaultOptions.js]: good call. changing this.

**Comment by lelandrichardson** [comment]
@lencioni @trotzig I've addressed most feedback. Think this is ready to merge now?

File: packages/happo-target-react-native/scripts/build-runner-android.sh (hunk 11 above)
Comment (lencioni) [inline packages/happo-target-react-native/scripts/build-runner-android.sh]: Where is the `cd`ing?

**Comment by lencioni** [comment]
I didn't look too closely at the updated commits, but if you feel good about it, go for it.

**Comment by trotzig** [comment]
I'm all 👍
